File: src/cli/tui/store.ts
/**
 * TUI store (GRACE full-screen interface).
 *
 * A single mutable state object + subscriber list. React components subscribe
 * via `useSyncExternalStore`; every mutation bumps `version` and notifies.
 * All interactive logic (input editing, history, scrolling, overlays) lives
 * here so it is pure, unit-testable and independent of the renderer.
 */
import type {
  ActivityItem,
  ActivityKind,
  FocusTarget,
  LoginField,
  LoginState,
  PaletteState,
  PermissionState,
  PickerKind,
  PickerOption,
  PickerState,
  SlashCommandDef,
  TuiInfo,
} from './types.ts';

/** Cap on activity lines — oldest lines drop to keep the TUI responsive. */
const MAX_ACTIVITY = 2_000;

let nextId = 1;

export class TuiStore {
  private listeners = new Set<() => void>();
  private versionValue = 0;

  // -- layout / mode ---------------------------------------------------------
  /** 'home' shows the branded startup screen; 'session' shows the activity log. */
  mode: 'home' | 'session' = 'home';
  focus: FocusTarget = 'input';

  // -- input -----------------------------------------------------------------
  input = '';
  cursor = 0;
  history: string[] = [];
  private historyIndex = -1;

  // -- activity --------------------------------------------------------------
  items: ActivityItem[] = [];
  /** True while an agent task is running. */
  busy = false;
  /** Lines scrolled up from the bottom; 0 = following the latest output. */
  scroll = 0;
  /** Files the running task has changed (real, from file-changed events). */
  changedFiles: string[] = [];
  private toolCalls = 0;

  // -- overlays --------------------------------------------------------------
  permission: PermissionState | null = null;
  picker: PickerState | null = null;
  palette: PaletteState | null = null;
  helpOpen = false;
  login: LoginState | null = null;

  // -- runtime facts ---------------------------------------------------------
  info: TuiInfo;

  /** Timestamp the current task started (real). */
  taskStartedAt: number | null = null;

  constructor(info: TuiInfo) {
    this.info = info;
  }

  // -------------------------------------------------------------------------
  // Subscription (React useSyncExternalStore)
  // -------------------------------------------------------------------------

  getVersion = (): number => {
    return this.versionValue;
  }

  subscribe = (listener: () => void): (() => void) => {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  };

  notify(): void {
    this.versionValue += 1;
    for (const l of this.listeners) l();
  }

  /** Force a redraw (Ctrl+R). */
  refresh(): void {
    this.notify();
  }

  // -------------------------------------------------------------------------
  // Input editing (real text editing — cursor, backspace, delete, history)
  // -------------------------------------------------------------------------

  insert(ch: string): void {
    if (!ch) return;
    this.input = this.input.slice(0, this.cursor) + ch + this.input.slice(this.cursor);
    this.cursor += ch.length;
    this.syncPalette();
    this.notify();
  }

  backspace(): void {
    if (this.cursor === 0) return;
    this.input = this.input.slice(0, this.cursor - 1) + this.input.slice(this.cursor);
    this.cursor -= 1;
    this.syncPalette();
    this.notify();
  }

  delete(): void {
    if (this.cursor >= this.input.length) return;
    this.input = this.input.slice(0, this.cursor) + this.input.slice(this.cursor + 1);
    this.syncPalette();
    this.notify();
  }

  moveLeft(): void {
    if (this.cursor > 0) {
      this.cursor -= 1;
      this.notify();
    }
  }

  moveRight(): void {
    if (this.cursor < this.input.length) {
      this.cursor += 1;
      this.notify();
    }
  }

  home(): void {
    if (this.cursor !== 0) {
      this.cursor = 0;
      this.notify();
    }
  }

  end(): void {
    if (this.cursor !== this.input.length) {
      this.cursor = this.input.length;
      this.notify();
    }
  }

  clearInput(): void {
    if (this.input === '' && this.palette === null) return;
    this.input = '';
    this.cursor = 0;
    this.closePalette();
    this.historyIndex = -1;
    this.notify();
  }

  setInput(text: string): void {
    this.input = text;
    this.cursor = text.length;
    this.historyIndex = -1;
    this.syncPalette();
    this.notify();
  }

  /** Commit the current line to history and reset the input. */
  submitInput(): void {
    const text = this.input.trim();
    if (text) {
      if (this.history[this.history.length - 1] !== text) this.history.push(text);
      if (this.history.length > 200) this.history.splice(0, this.history.length - 200);
    }
    this.historyIndex = -1;
    this.input = '';
    this.cursor = 0;
    this.closePalette();
    this.notify();
  }

  historyUp(): void {
    if (this.history.length === 0) return;
    if (this.historyIndex === -1) {
      this.historyIndex = this.history.length - 1;
    } else if (this.historyIndex > 0) {
      this.historyIndex -= 1;
    } else {
      return;
    }
    this.input = this.history[this.historyIndex] ?? '';
    this.cursor = this.input.length;
    this.syncPalette();
    this.notify();
  }

  historyDown(): void {
    if (this.historyIndex === -1) return;
    this.historyIndex += 1;
    if (this.historyIndex >= this.history.length) {
      this.historyIndex = -1;
      this.input = '';
    } else {
      this.input = this.history[this.historyIndex] ?? '';
    }
    this.cursor = this.input.length;
    this.syncPalette();
    this.notify();
  }

  // -------------------------------------------------------------------------
  // Activity feed + scrolling
  // -------------------------------------------------------------------------

  /**
   * Push one or more activity lines. Multi-line text is split into individual
   * lines so the scroll math and rendering stay line-based.
   */
  push(kind: ActivityKind, text: string): void {
    if (!text) return;
    const lines = text.split('\n');
    for (const raw of lines) {
      const line = raw.replace(/\r/g, '').trimEnd();
      if (line === '') continue;
      this.items.push({ id: nextId++, kind, text: line });
    }
    if (this.items.length > MAX_ACTIVITY) {
      this.items.splice(0, this.items.length - MAX_ACTIVITY);
    }
    // Following the bottom stays at the bottom automatically (scroll 0).
    this.notify();
  }

  clearActivity(): void {
    this.items = [];
    this.scroll = 0;
    this.changedFiles = [];
    this.toolCalls = 0;
    this.mode = 'home';
    this.notify();
  }

  scrollUp(lines: number): void {
    const maxScroll = this.items.length - 1;
    if (maxScroll <= 0) return;
    this.scroll = Math.min(maxScroll, this.scroll + lines);
    this.notify();
  }

  scrollDown(lines: number): void {
    this.scroll = Math.max(0, this.scroll - lines);
    this.notify();
  }

  scrollToBottom(): void {
    if (this.scroll !== 0) {
      this.scroll = 0;
      this.notify();
    }
  }

  toggleFocus(): void {
    this.focus = this.focus === 'input' ? 'activity' : 'input';
    this.notify();
  }

  // -------------------------------------------------------------------------
  // Task lifecycle
  // -------------------------------------------------------------------------

  setBusy(busy: boolean): void {
    this.busy = busy;
    if (busy) this.taskStartedAt = Date.now();
    else this.taskStartedAt = null;
    this.notify();
  }

  recordToolCall(): void {
    this.toolCalls += 1;
  }

  get toolCallCount(): number {
    return this.toolCalls;
  }

  recordChangedFile(path: string): void {
    if (!this.changedFiles.includes(path)) this.changedFiles.push(path);
  }

  // -------------------------------------------------------------------------
  // Permission dialog
  // -------------------------------------------------------------------------

  /** Open the permission dialog; resolves when the user answers y/n/a. */
  askPermission(command: string, reasons: string[]): Promise<boolean> {
    return new Promise((resolve) => {
      this.permission = { id: nextId++, command, reasons, resolve };
      this.notify();
    });
  }

  answerPermission(allowed: boolean): void {
    const p = this.permission;
    if (!p) return;
    this.permission = null;
    p.resolve(allowed);
    this.notify();
  }

  // -------------------------------------------------------------------------
  // Pickers (model / provider)
  // -------------------------------------------------------------------------

  openPicker(kind: PickerKind, title: string, options: PickerOption[], onSelect: (opt: PickerOption, index: number) => void, onCancel: () => void): void {
    this.picker = { id: nextId++, title, options, all: options, filter: '', selected: Math.max(0, options.findIndex((o) => o.current)), onSelect, onCancel };
    this.notify();
  }

  pickerFilter(query: string): void {
    const p = this.picker;
    if (!p) return;
    p.filter = query;
    p.options = p.all.filter((o) => o.label.toLowerCase().includes(query.toLowerCase()));
    if (p.selected >= p.options.length) p.selected = Math.max(0, p.options.length - 1);
    this.notify();
  }

  pickerMove(delta: number): void {
    const p = this.picker;
    if (!p || p.options.length === 0) return;
    p.selected = (p.selected + delta + p.options.length) % p.options.length;
    this.notify();
  }

  pickerSelect(): void {
    const p = this.picker;
    if (!p) return;
    const opt = p.options[p.selected];
    if (!opt || opt.disabled) return;
    const index = p.selected;
    this.picker = null;
    p.onSelect(opt, index);
  }

  closePicker(): void {
    const p = this.picker;
    this.picker = null;
    p?.onCancel();
    this.notify();
  }

  // -------------------------------------------------------------------------
  // Command palette
  // -------------------------------------------------------------------------

  /** Open (or keep) the palette for a slash-prefixed input. */
  syncPalette(): void {
    if (this.input.startsWith('/') && this.input.length >= 1) {
      if (!this.palette) {
        this.palette = { commands: [], query: '', selected: 0 };
        this.notify();
      }
      this.palette.query = this.input;
    } else {
      this.closePalette();
    }
  }

  /** Refresh palette entries (called with the real command table). */
  setPaletteCommands(commands: SlashCommandDef[]): void {
    if (!this.palette) return;
    this.palette.commands = commands;
    this.notify();
  }

  paletteMove(delta: number): void {
    const p = this.palette;
    if (!p || p.commands.length === 0) return;
    p.selected = (p.selected + delta + p.commands.length) % p.commands.length;
    this.notify();
  }

  /**
   * Visible palette rows for the current query. Matches on the FIRST token so
   * "/model groq" still highlights /model while keeping typed args intact.
   */
  paletteRows(): SlashCommandDef[] {
    if (!this.palette) return [];
    const raw = this.palette.query.startsWith('/') ? this.palette.query.slice(1) : this.palette.query;
    const query = (raw.split(/\s+/)[0] ?? '').toLowerCase();
    const rows = query
      ? this.palette.commands.filter((c) => c.name.toLowerCase().startsWith('/' + query))
      : this.palette.commands;
    return rows.slice(0, 10);
  }

  closePalette(): void {
    if (this.palette) {
      this.palette = null;
      this.notify();
    }
  }

  // -------------------------------------------------------------------------
  // Help overlay
  // -------------------------------------------------------------------------

  openHelp(): void {
    this.helpOpen = true;
    this.notify();
  }

  closeHelp(): void {
    this.helpOpen = false;
    this.notify();
  }

  // -------------------------------------------------------------------------
  // Login/register overlay
  // -------------------------------------------------------------------------

  openLogin(purpose: 'login' | 'register', emailArg: string): void {
    this.login = {
      purpose,
      email: emailArg.trim() ?? '',
      field: 'email',
      password: '',
      confirm: '',
      busy: false,
    };
    this.notify();
  }

  loginType(ch: string): void {
    const l = this.login;
    if (!l || l.busy) return;
    const field: LoginField = l.field;
    if (field === 'email') l.email += ch;
    else if (field === 'password') l.password += ch;
    else l.confirm += ch;
    this.notify();
  }

  loginBackspace(): void {
    const l = this.login;
    if (!l || l.busy) return;
    if (l.field === 'email') l.email = l.email.slice(0, -1);
    else if (l.field === 'password') l.password = l.password.slice(0, -1);
    else l.confirm = l.confirm.slice(0, -1);
    this.notify();
  }

  loginNextField(): void {
    const l = this.login;
    if (!l || l.busy) return;
    if (l.field === 'email') l.field = 'password';
    else if (l.field === 'password') l.field = l.purpose === 'register' ? 'confirm' : 'email';
    else l.field = 'email';
    this.notify();
  }

  loginSetField(field: LoginField): void {
    const l = this.login;
    if (!l) return;
    l.field = field;
    this.notify();
  }

  loginError(error: string): void {
    const l = this.login;
    if (!l) return;
    l.error = error;
    l.busy = false;
    this.notify();
  }

  loginBusy(): void {
    const l = this.login;
    if (!l) return;
    l.busy = true;
    l.error = undefined;
    this.notify();
  }

  closeLogin(): void {
    this.login = null;
    this.notify();
  }
}

File: grace/cli/tui/store.py
"""TUI store (port of src/cli/tui/store.ts).

A single mutable state object + subscriber list. Every mutation bumps
`version` and notifies. All interactive logic (input editing, history,
scrolling, overlays) lives here so it is pure, unit-testable and independent
of the renderer.
"""

import threading
import time

from grace.cli.tui.commands import HOME_SHORTCUTS, SLASH_COMMANDS

MAX_ACTIVITY = 2_000

# NOTE: keep the counter under a DIFFERENT name than the function — the
# `def` would otherwise shadow the list, making every call raise
# "'function' object is not subscriptable" (TS→Python migration trap).
_id_counter = [1]


def _next_id() -> int:
    _id_counter[0] += 1
    return _id_counter[0]


class TuiStore:
    def __init__(self, info: dict) -> None:
        self.info = info
        self.listeners: set = set()
        self.version_value = 0

        # -- layout / mode -------------------------------------------------
        self.mode = "home"  # 'home' | 'session'
        self.focus = "input"
        self.shortcut_index = 0

        # -- input ---------------------------------------------------------
        self.input = ""
        self.cursor = 0
        self.history: list[str] = []
        self.history_index = -1

        # -- activity ------------------------------------------------------
        self.items: list[dict] = []
        self.busy = False
        self.scroll = 0
        self.changed_files: list[str] = []
        self._tool_calls = 0

        # -- overlays ------------------------------------------------------
        self.permission: dict | None = None
        self.picker: dict | None = None
        self.palette: dict | None = None
        self.help_open = False
        self.login: dict | None = None

        self.task_started_at: float | None = None

    # ---------------------------------------------------------------------
    # Subscription
    # ---------------------------------------------------------------------

    def get_version(self) -> int:
        return self.version_value

    def subscribe(self, listener) -> callable:
        self.listeners.add(listener)
        return lambda: self.listeners.discard(listener)

    def notify(self) -> None:
        self.version_value += 1
        for l in list(self.listeners):
            l()

    def refresh(self) -> None:
        self.notify()

    # ---------------------------------------------------------------------
    # Input editing
    # ---------------------------------------------------------------------

    def insert(self, ch: str) -> None:
        if not ch:
            return
        self.input = self.input[:self.cursor] + ch + self.input[self.cursor:]
        self.cursor += len(ch)
        self.sync_palette()
        self.notify()

    def backspace(self) -> None:
        if self.cursor == 0:
            return
        self.input = self.input[:self.cursor - 1] + self.input[self.cursor:]
        self.cursor -= 1
        self.sync_palette()
        self.notify()

    def delete(self) -> None:
        if self.cursor >= len(self.input):
            return
        self.input = self.input[:self.cursor] + self.input[self.cursor + 1:]
        self.sync_palette()
        self.notify()

    def move_left(self) -> None:
        if self.cursor > 0:
            self.cursor -= 1
            self.notify()

    def move_right(self) -> None:
        if self.cursor < len(self.input):
            self.cursor += 1
            self.notify()

    def home(self) -> None:
        if self.cursor != 0:
            self.cursor = 0
            self.notify()

    def end(self) -> None:
        if self.cursor != len(self.input):
            self.cursor = len(self.input)
            self.notify()

    def clear_input(self) -> None:
        if self.input == "" and self.palette is None:
            return
        self.input = ""
        self.cursor = 0
        self.close_palette()
        self.history_index = -1
        self.notify()

    def set_input(self, text: str) -> None:
        self.input = text
        self.cursor = len(text)
        self.history_index = -1
        self.sync_palette()
        self.notify()

    def submit_input(self) -> None:
        text = self.input.strip()
        if text:
            if not self.history or self.history[-1] != text:
                self.history.append(text)
            if len(self.history) > 200:
                self.history = self.history[-200:]
        self.history_index = -1
        self.input = ""
        self.cursor = 0
        self.focus = "input"
        self.close_palette()
        self.notify()

    def history_up(self) -> None:
        if not self.history:
            return
        if self.history_index == -1:
            self.history_index = len(self.history) - 1
        elif self.history_index > 0:
            self.history_index -= 1
        else:
            return
        self.input = self.history[self.history_index] or ""
        self.cursor = len(self.input)
        self.sync_palette()
        self.notify()

    def history_down(self) -> None:
        if self.history_index == -1:
            return
        self.history_index += 1
        if self.history_index >= len(self.history):
            self.history_index = -1
            self.input = ""
        else:
            self.input = self.history[self.history_index] or ""
        self.cursor = len(self.input)
        self.sync_palette()
        self.notify()

    # ---------------------------------------------------------------------
    # Activity feed + scrolling
    # ---------------------------------------------------------------------

    def push(self, kind: str, text: str) -> None:
        if not text:
            return
        lines = text.split("\n")
        for raw in lines:
            line = raw.replace("\r", "").rstrip()
            if line == "":
                continue
            self.items.append({"id": _next_id(), "kind": kind, "text": line})
        if len(self.items) > MAX_ACTIVITY:
            self.items = self.items[-MAX_ACTIVITY:]
        self.notify()

    def clear_activity(self) -> None:
        self.items = []
        self.scroll = 0
        self.changed_files = []
        self._tool_calls = 0
        self.mode = "home"
        self.notify()

    def scroll_up(self, lines: int) -> None:
        max_scroll = len(self.items) - 1
        if max_scroll <= 0:
            return
        self.scroll = min(max_scroll, self.scroll + lines)
        self.notify()

    def scroll_down(self, lines: int) -> None:
        self.scroll = max(0, self.scroll - lines)
        self.notify()

    def scroll_to_bottom(self) -> None:
        if self.scroll != 0:
            self.scroll = 0
            self.notify()

    def toggle_focus(self) -> None:
        if self.mode == "home":
            self.focus = "shortcuts" if self.focus == "input" else "input"
        else:
            self.focus = "activity" if self.focus == "input" else "input"
        self.notify()

    def set_focus(self, focus: str) -> None:
        if self.focus != focus:
            self.focus = focus
            self.notify()

    def shortcut_move(self, delta: int) -> None:
        if not HOME_SHORTCUTS:
            return
        self.shortcut_index = (self.shortcut_index + delta + len(HOME_SHORTCUTS)) % len(HOME_SHORTCUTS)
        self.notify()

    # ---------------------------------------------------------------------
    # Task lifecycle
    # ---------------------------------------------------------------------

    def set_busy(self, busy: bool) -> None:
        self.busy = busy
        self.task_started_at = time.monotonic() if busy else None
        self.notify()

    def record_tool_call(self) -> None:
        self._tool_calls += 1

    @property
    def tool_call_count(self) -> int:
        return self._tool_calls

    def record_changed_file(self, path: str) -> None:
        if path not in self.changed_files:
            self.changed_files.append(path)

    # ---------------------------------------------------------------------
    # Permission dialog
    # ---------------------------------------------------------------------

    def ask_permission(self, command: str, reasons: list[str]) -> bool:
        """Open the permission dialog; blocks (in the worker thread) until the
        user answers y/n/a."""
        holder: dict = {"result": False, "done": threading.Event()}
        self.permission = {"id": _next_id(), "command": command, "reasons": reasons, "holder": holder}
        self.notify()
        holder["done"].wait()
        return holder["result"]

    def answer_permission(self, allowed: bool) -> None:
        p = self.permission
        if not p:
            return
        self.permission = None
        p["holder"]["result"] = allowed
        p["holder"]["done"].set()
        self.notify()

    # ---------------------------------------------------------------------
    # Pickers (model / provider)
    # ---------------------------------------------------------------------

    def open_picker(self, kind: str, title: str, options: list[dict], on_select, on_cancel=None) -> None:
        current = 0
        for i, o in enumerate(options):
            if o.get("current"):
                current = i
                break
        self.picker = {
            "id": _next_id(),
            "kind": kind,
            "title": title,
            "options": list(options),
            "all": list(options),
            "filter": "",
            "selected": max(0, current),
            "onSelect": on_select,
            "onCancel": on_cancel or (lambda: None),
        }
        self.notify()

    def picker_filter(self, query: str) -> None:
        p = self.picker
        if not p:
            return
        p["filter"] = query
        p["options"] = [o for o in p["all"] if query.lower() in o.get("label", "").lower()]
        if p["selected"] >= len(p["options"]):
            p["selected"] = max(0, len(p["options"]) - 1)
        self.notify()

    def picker_move(self, delta: int) -> None:
        p = self.picker
        if not p or not p["options"]:
            return
        p["selected"] = (p["selected"] + delta + len(p["options"])) % len(p["options"])
        self.notify()

    def picker_select(self) -> None:
        p = self.picker
        if not p:
            return
        opts = p["options"]
        if p["selected"] >= len(opts):
            return
        opt = opts[p["selected"]]
        if opt.get("disabled"):
            return
        index = p["selected"]
        on_select = p["onSelect"]
        self.picker = None
        on_select(opt, index)

    def close_picker(self) -> None:
        p = self.picker
        self.picker = None
        if p and p.get("onCancel"):
            p["onCancel"]()
        self.notify()

    # ---------------------------------------------------------------------
    # Command palette
    # ---------------------------------------------------------------------

    def sync_palette(self) -> None:
        if self.input.startswith("/") and len(self.input) >= 1:
            if not self.palette:
                self.palette = {"commands": list(SLASH_COMMANDS), "query": "", "selected": 0}
                self.notify()
            self.palette["query"] = self.input
        else:
            self.close_palette()

    def set_palette_commands(self, commands: list[dict]) -> None:
        if not self.palette:
            return
        self.palette["commands"] = commands
        self.notify()

    def palette_move(self, delta: int) -> None:
        p = self.palette
        if not p or not p["commands"]:
            return
        p["selected"] = (p["selected"] + delta + len(p["commands"])) % len(p["commands"])
        self.notify()

    def palette_rows(self) -> list[dict]:
        """Visible palette rows for the current query. Matches on the FIRST
        token so "/model groq" still highlights /model while keeping typed args
        intact."""
        if not self.palette:
            return []
        raw = self.palette["query"][1:] if self.palette["query"].startswith("/") else self.palette["query"]
        query = (raw.split()[0] if raw.split() else "").lower()
        rows = [c for c in self.palette["commands"] if c["name"].lower().startswith("/" + query)] if query else self.palette["commands"]
        return rows[:10]

    def close_palette(self) -> None:
        if self.palette:
            self.palette = None
            self.notify()

    # ---------------------------------------------------------------------
    # Help overlay
    # ---------------------------------------------------------------------

    def open_help(self) -> None:
        self.help_open = True
        self.notify()

    def close_help(self) -> None:
        self.help_open = False
        self.notify()

    # ---------------------------------------------------------------------
    # Login/register overlay
    # ---------------------------------------------------------------------

    def open_login(self, purpose: str, email_arg: str) -> None:
        self.login = {"purpose": purpose, "email": (email_arg or "").strip(), "field": "email", "password": "", "confirm": "", "busy": False}
        self.notify()

    def login_type(self, ch: str) -> None:
        l = self.login
        if not l or l.get("busy"):
            return
        field = l["field"]
        if field == "email":
            l["email"] += ch
        elif field == "password":
            l["password"] += ch
        else:
            l["confirm"] += ch
        self.notify()

    def login_backspace(self) -> None:
        l = self.login
        if not l or l.get("busy"):
            return
        field = l["field"]
        if field == "email":
            l["email"] = l["email"][:-1]
        elif field == "password":
            l["password"] = l["password"][:-1]
        else:
            l["confirm"] = l["confirm"][:-1]
        self.notify()

    def login_next_field(self) -> None:
        l = self.login
        if not l or l.get("busy"):
            return
        if l["field"] == "email":
            l["field"] = "password"
        elif l["field"] == "password":
            l["field"] = "confirm" if l["purpose"] == "register" else "email"
        else:
            l["field"] = "email"
        self.notify()

    def login_set_field(self, field: str) -> None:
        l = self.login
        if not l:
            return
        l["field"] = field
        self.notify()

    def login_error(self, error: str) -> None:
        l = self.login
        if not l:
            return
        l["error"] = error
        l["busy"] = False
        self.notify()

    def login_busy(self) -> None:
        l = self.login
        if not l:
            return
        l["busy"] = True
        l.pop("error", None)
        self.notify()

    def close_login(self) -> None:
        self.login = None
        self.notify()
